### PR TITLE
Add WebSockets connection pool limit when shields are enabled.

### PR DESCRIPTION
**Pull request by goodov**
Resolves https://github.com/brave/brave-browser/issues/19990

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by iefremov** [comment]
Most probably we don't need a formal sec review, but I'd double-check with someone from the sec team. Maybe @fmarier you could take a quick look?

File: chromium_src/third_party/blink/common/features.cc [hunk 1]
```diff
@@ -40,5 +40,9 @@ const base::Feature kFileSystemAccessAPI{"FileSystemAccessAPI",
 const base::Feature kPartitionBlinkMemoryCache{
     "PartitionBlinkMemoryCache", base::FEATURE_DISABLED_BY_DEFAULT};
 
+// Enable WebSockets connection pool limit per eTLD+1 for each renderer.
+const base::Feature kRestrictWebSocketsPool{"RestrictWebSocketsPool",
+                                            base::FEATURE_DISABLED_BY_DEFAULT};
```
Comment (iefremov) [inline chromium_src/third_party/blink/common/features.cc:45]: perhaps we can enable by default, the change doesn't look very dangerous?

File: third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (iefremov) [inline third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.h:1]: is almost 2022! (in all files)

File: chromium_src/third_party/blink/common/features.cc (hunk 1 above)
Comment (goodov) [reply chromium_src/third_party/blink/common/features.cc:45]: 👌 I'll make it enabled by default and will add a `brave://flag` to be safe.

**Comment by iefremov** [comment]
perhaps a browser test?

**Comment by goodov** [comment]
> perhaps a browser test?

of course. currently working on it.

**Review by fmarier — APPROVED** [review]
>Most probably we don't need a formal sec review, but I'd double-check with someone from the sec team.

The high-level approach sounds good to me. It's not adding (or really modifying) any new networks requests, so I don't think we need to do a full sec review.

I assume @pes10k has confirmed that this mitigation is effective against the pool party attack?

**Comment by pes10k** [comment]
Yep! Limiting each eTLD+1 to a maximum of 10 websockets would effectively defeat the attack (an attacker could hypothetically still carry out the attack by controlling ~20 domains at the same time, and coordinating between them, but this approach makes the attack sufficiently difficult to carry out that I think we can consider it solved (at least when carried out through the WebSocket pool, which is the most concerning example).

Just to confirm @goodov, this PR does

> Partition: impose a per-site (as determined by top-level frame eTLD+1) cap of how many websockets can be open. Initial suggestion is 10, but any cap thats << the global cap would work

but not 

> De-limit: remove, or greatly expand, the global cap, to increase the number of eTLD+1s an attack has to control to carry out the attack

Is that correct?

**Comment by goodov** [comment]
> Just to confirm @goodov, this PR does
> 
> > Partition: impose a per-site (as determined by top-level frame eTLD+1) cap of how many websockets can be open. Initial suggestion is 10, but any cap thats << the global cap would work
> 
> but not
> 
> > De-limit: remove, or greatly expand, the global cap, to increase the number of eTLD+1s an attack has to control to carry out the attack
> 
> Is that correct?

yes, that is correct!

File: third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  DCHECK(context);
+  if (auto* window = DynamicTo<LocalDOMWindow>(context)) {
+    auto* frame = window->GetFrame();
+    if (!frame)
+      frame = window->GetDisconnectedFrame();
+    if (frame) {
+      if (auto* top_frame_security_context =
+              frame->Top()->GetSecurityContext()) {
+        return top_frame_security_context->GetSecurityOrigin()
+            ->GetOriginOrPrecursorOriginIfOpaque();
```
Comment (goodov) [inline third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc:32]: see `SandboxedFramesAreLimited` test.

**Review by iefremov — APPROVED** [review]

**Comment by goodov** [comment]
@bridiver `chromium_src`, `patches` review ping, other reviewers are currently on PTO.

**Comment by iefremov** [comment]
@mkarolin @bridiver PTAL, thanks!

File: chromium_src/third_party/blink/renderer/modules/websockets/websocket_channel_impl.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+#include "src/third_party/blink/renderer/modules/websockets/websocket_channel_impl.h"
+
+#undef ShouldDisallowConnection
+#undef TearDownFailedConnection
+#undef WebSocketChannelImpl
+
+namespace blink {
+
+class MODULES_EXPORT WebSocketChannelImpl
```
Comment (mkarolin) [inline chromium_src/third_party/blink/renderer/modules/websockets/websocket_channel_impl.h:33]: Should we restore `final`ity here?

File: third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  class CORE_EXPORT ResourceInUseTracker {
+   public:
+    explicit ResourceInUseTracker(String resource_id_in_use);
+    ~ResourceInUseTracker();
+
+    const String& resource_id_in_use() const { return resource_id_in_use_; }
+
+   private:
+    String resource_id_in_use_;
```
Comment (mkarolin) [inline third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.h:36]: Nit: seems like this and the getter could just be called resource_id_/resource_id()?

File: third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+std::unique_ptr<ResourcePoolLimiter::ResourceInUseTracker>
+ResourcePoolLimiter::IssueResourceInUseTracker(ExecutionContext* context,
+                                               ResourceType resource_type) {
+  DCHECK(context);
+  String resource_id = GetResourceIdInUse(
+      GetTopFrameOrContextSecurityOrigin(context), resource_type);
+
+  MutexLocker locker(resources_in_use_lock_);
+  // `insert` doesn't change the value if it already exists.
+  int& resource_in_use_value =
```
Comment (mkarolin) [inline third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc:91]: Nit: resource_in_use_count might be a bit more clear.

File: third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  if (origin_id.IsEmpty())
+    origin_id = origin->Host();
+  string_builder.Append(origin_id);
+  return string_builder.ToString();
+}
+
+int GetResourceLimit(ResourceType resource_type) {
+  switch (resource_type) {
+    case ResourceType::kWebSocket:
+      return 10;
```
Comment (mkarolin) [inline third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc:54]: Should this be hard-coded? Would it make sense to make it a feature param, for example, and control it via Griffin?

**Comment by iefremov** [comment]
@mkarolin turned out we want to merge it asap, would you mind approving it and then we can do a follow-up?

**Review by mkarolin — APPROVED** [review]
++ with a couple of nits and questions that can be address in a followup.

**Comment by mkarolin** [comment]
> @mkarolin turned out we want to merge it asap, would you mind approving it and then we can do a follow-up?

No problem.

**Pull request merged**